File: packages/analysis/src/stemmer-nl.ts
const A = new Set(["a", "à", "á", "â", "ä"]);
const E = new Set(["e", "è", "é", "ê", "ë"]);
const I = new Set(["i", "ì", "í", "î", "ï"]);
const O = new Set(["o", "ò", "ó", "ô", "ö"]);
const U = new Set(["u", "ù", "ú", "û", "ü"]);
const AEIOU = new Set([...A, ...E, ...I, ...O, ...U]);
const AIOU = new Set([...A, ...I, ...O, ...U]);
const VOWELS = new Set([...AEIOU, "y"]);

function vowelLengthAt(word: string, index: number): number {
  if (word.slice(index, index + 2) === "ij") return 2;
  return VOWELS.has(word[index] ?? "") ? 1 : 0;
}

function measure(word: string): { p1: number; p2: number } {
  const regions: number[] = [];
  let index = 0;
  while (regions.length < 2 && index < word.length) {
    while (index < word.length && vowelLengthAt(word, index) === 0) index++;
    let vowels = 0;
    while (index < word.length) {
      const length = vowelLengthAt(word, index);
      if (length === 0) break;
      vowels++;
      index += length;
    }
    if (vowels === 0 || index >= word.length) break;
    index++;
    regions.push(index);
  }
  return { p1: regions[0] ?? word.length, p2: regions[1] ?? word.length };
}

function isC(prefix: string): boolean {
  if (prefix.length === 0 || prefix.endsWith("ij")) return false;
  return !VOWELS.has(prefix.at(-1) as string);
}

function isV(prefix: string): boolean {
  return prefix.endsWith("ij") || VOWELS.has(prefix.at(-1) ?? "");
}

function inRegion(word: string, suffix: string, region: number): boolean {
  return word.length - suffix.length >= region;
}

function longestSuffix(
  word: string,
  suffixes: readonly string[],
): string | undefined {
  let best: string | undefined;
  for (const suffix of suffixes) {
    if (
      word.endsWith(suffix) &&
      (best === undefined || suffix.length > best.length)
    ) {
      best = suffix;
    }
  }
  return best;
}

function lengthenV(word: string): string {
  if (word.length < 2) return word;
  const last = word.at(-1) as string;
  if (VOWELS.has(last) || last === "w" || last === "x") return word;
  const prefix = word.slice(0, -1);
  if (prefix.endsWith("eë")) return `${prefix.slice(0, -2)}eëe${last}`;
  if (prefix.endsWith("ië")) return `${prefix.slice(0, -2)}iee${last}`;

  const vowel = prefix.at(-1);
  if (vowel === undefined) return word;
  const before = prefix.slice(0, -1);
  if (A.has(vowel) || O.has(vowel) || U.has(vowel)) {
    if (before.length === 0 || !AEIOU.has(before.at(-1) as string)) {
      return `${prefix}${vowel}${last}`;
    }
  } else if (E.has(vowel) && vowel !== "ë") {
    if (before.length > 0 && AEIOU.has(before.at(-1) as string)) return word;
    const base = before.length === 0 ? before : before.slice(0, -1);
    if (AIOU.has(base.at(-1) ?? "")) return word;
    if (base.length === 1 && E.has(base)) return word;
    const skipped = base.slice(0, -1);
    if (
      skipped.length >= 2 &&
      AIOU.has(skipped.at(-1) as string) &&
      !AEIOU.has(skipped.at(-2) as string)
    ) {
      return word;
    }
    return `${prefix}${vowel}${last}`;
  }
  return word;
}

function step1(word: string, p1: number): { word: string; changed: boolean } {
  const suffix = longestSuffix(word, [
    "'s",
    "ies",
    "és",
    "aus",
    "nde",
    "en",
    "es",
    "s",
  ]);
  if (suffix === undefined) return { word, changed: false };
  const prefix = word.slice(0, -suffix.length);
  switch (suffix) {
    case "'s":
      return { word: prefix, changed: true };
    case "s":
      if (!inRegion(word, suffix, p1) || !isC(prefix))
        return { word, changed: false };
      if (prefix.endsWith("t") && prefix.length - 1 >= p1)
        return { word, changed: false };
      return { word: prefix, changed: true };
    case "ies":
      return inRegion(word, suffix, p1)
        ? { word: `${prefix}ie`, changed: true }
        : { word, changed: false };
    case "es": {
      if (
        prefix.endsWith("ar") &&
        prefix.length - 2 >= p1 &&
        isC(prefix.slice(0, -2))
      ) {
        return { word: lengthenV(prefix), changed: true };
      }
      if (
        prefix.endsWith("er") &&
        prefix.length - 2 >= p1 &&
        isC(prefix.slice(0, -2))
      ) {
        return { word: prefix, changed: true };
      }
      return inRegion(word, suffix, p1) && isC(prefix)
        ? { word: `${prefix}e`, changed: true }
        : { word, changed: false };
    }
    case "és":
      return inRegion(word, suffix, p1)
        ? { word: `${prefix}é`, changed: true }
        : { word, changed: false };
    case "aus":
      return inRegion(word, suffix, p1) && isV(prefix)
        ? { word: `${prefix}au`, changed: true }
        : { word, changed: false };
    case "en": {
      if (prefix.endsWith("hed") && prefix.length - 3 >= p1) {
        return { word: `${prefix.slice(0, -3)}heid`, changed: true };
      }
      if (prefix.endsWith("nd")) return { word: prefix, changed: true };
      if (
        prefix.endsWith("d") &&
        prefix.length - 1 >= p1 &&
        isC(prefix.slice(0, -1))
      ) {
        return { word: prefix.slice(0, -1), changed: true };
      }
      if (
        (prefix.endsWith("i") || prefix.endsWith("j")) &&
        isV(prefix.slice(0, -1))
      ) {
        return { word: prefix, changed: true };
      }
      return inRegion(word, suffix, p1) && isC(prefix)
        ? { word: lengthenV(prefix), changed: true }
        : { word, changed: false };
    }
    case "nde":
      return { word: `${prefix}nd`, changed: true };
  }
  return { word, changed: false };
}

function step2(word: string, p1: number): { word: string; changed: boolean } {
  const suffix = longestSuffix(word, [
    "lijke",
    "ische",
    "ieve",
    "ene",
    "je",
    "ge",
    "de",
    "te",
    "se",
    "re",
    "le",
  ]);
  if (suffix === undefined) return { word, changed: false };
  let prefix = word.slice(0, -suffix.length);
  if (suffix === "je") {
    if (prefix.endsWith("'t"))
      return { word: prefix.slice(0, -2), changed: true };
    if (
      prefix.endsWith("et") &&
      prefix.length - 2 >= p1 &&
      isC(prefix.slice(0, -2))
    ) {
      return { word: prefix.slice(0, -2), changed: true };
    }
    if (prefix.endsWith("rnt"))
      return { word: `${prefix.slice(0, -3)}rn`, changed: true };
    if (
      prefix.endsWith("t") &&
      prefix.length - 1 >= p1 &&
      isV(prefix.slice(0, -2))
    ) {
      return { word: prefix.slice(0, -1), changed: true };
    }
    if (prefix.endsWith("ink"))
      return { word: `${prefix.slice(0, -3)}ing`, changed: true };
    if (prefix.endsWith("mp"))
      return { word: `${prefix.slice(0, -2)}m`, changed: true };
    if (prefix.endsWith("'") && prefix.length - 1 >= p1)
      return { word: prefix.slice(0, -1), changed: true };
    return inRegion(word, suffix, p1) && isC(prefix)
      ? { word: prefix, changed: true }
      : { word, changed: false };
  }
  if (!inRegion(word, suffix, p1)) return { word, changed: false };
  switch (suffix) {
    case "ge":
      return { word: `${prefix}g`, changed: true };
    case "lijke":
      return { word: `${prefix}lijk`, changed: true };
    case "ische":
      return { word: `${prefix}isch`, changed: true };
    case "de":
      return isC(prefix)
        ? { word: prefix, changed: true }
        : { word, changed: false };
    case "te":
      return { word: `${prefix}t`, changed: true };
    case "se":
      return { word: `${prefix}s`, changed: true };
    case "re":
      return { word: `${prefix}r`, changed: true };
    case "le":
      prefix += "l";
      return { word: lengthenV(prefix), changed: true };
    case "ene":
      if (!isC(prefix)) return { word, changed: false };
      return { word: lengthenV(`${prefix}en`), changed: true };
    case "ieve":
      return isC(prefix)
        ? { word: `${prefix}ief`, changed: true }
        : { word, changed: false };
  }
  return { word, changed: false };
}

function step3(
  word: string,
  p1: number,
  p2: number,
): { word: string; changed: boolean } {
  const suffix = longestSuffix(word, [
    "iteit",
    "isme",
    "erij",
    "arij",
    "heid",
    "ster",
    "rder",
    "atie",
    "ing",
    "sel",
    "fie",
    "gie",
    "tst",
    "dst",
  ]);
  if (suffix === undefined) return { word, changed: false };
  const prefix = word.slice(0, -suffix.length);
  switch (suffix) {
    case "atie":
      return inRegion(word, suffix, p1)
        ? { word: `${prefix}eer`, changed: true }
        : { word, changed: false };
    case "iteit":
      return inRegion(word, suffix, p1)
        ? { word: lengthenV(prefix), changed: true }
        : { word, changed: false };
    case "heid":
    case "sel":
    case "ster":
      return inRegion(word, suffix, p1)
        ? { word: prefix, changed: true }
        : { word, changed: false };
    case "rder":
      return { word: `${prefix}r`, changed: true };
    case "ing":
    case "isme":
    case "erij":
      if (prefix.endsWith("ild")) return { word: `${prefix}er`, changed: true };
      return inRegion(word, suffix, p1)
        ? { word: lengthenV(prefix), changed: true }
        : { word, changed: false };
    case "arij":
      return inRegion(word, suffix, p1) && isC(prefix)
        ? { word: `${prefix}aar`, changed: true }
        : { word, changed: false };
    case "fie":
      return inRegion(word, suffix, p2)
        ? { word: lengthenV(`${prefix}f`), changed: true }
        : { word, changed: false };
    case "gie":
      return inRegion(word, suffix, p2)
        ? { word: lengthenV(`${prefix}g`), changed: true }
        : { word, changed: false };
    case "tst":
      return inRegion(word, suffix, p1) && isC(prefix)
        ? { word: `${prefix}t`, changed: true }
        : { word, changed: false };
    case "dst":
      return inRegion(word, suffix, p1) && isC(prefix)
        ? { word: `${prefix}d`, changed: true }
        : { word, changed: false };
  }
  return { word, changed: false };
}

function step4(word: string, p1: number): { word: string; changed: boolean } {
  const first = longestSuffix(word, [
    "achtiger",
    "achtigst",
    "eriger",
    "erigst",
    "ioneel",
    "lijker",
    "lijkst",
    "achtig",
    "atief",
    "baar",
    "naar",
    "laar",
    "raar",
    "tant",
    "erig",
    "end",
  ]);
  if (first !== undefined && inRegion(word, first, p1)) {
    const prefix = word.slice(0, -first.length);
    switch (first) {
      case "ioneel":
        return { word: `${prefix}ie`, changed: true };
      case "atief":
        return { word: `${prefix}eer`, changed: true };
      case "baar":
        return { word: prefix, changed: true };
      case "naar":
        if (isV(prefix)) return { word: `${prefix}n`, changed: true };
        break;
      case "laar":
        if (isV(prefix)) return { word: `${prefix}l`, changed: true };
        break;
      case "raar":
        if (isV(prefix)) return { word: `${prefix}r`, changed: true };
        break;
      case "tant":
        return { word: `${prefix}teer`, changed: true };
      case "lijker":
      case "lijkst":
        return { word: `${prefix}lijk`, changed: true };
      case "achtig":
      case "achtiger":
      case "achtigst":
        return { word: prefix, changed: true };
      case "erig":
      case "eriger":
      case "erigst":
      case "end":
        if (isC(prefix)) return { word: lengthenV(prefix), changed: true };
        break;
    }
  }
  const second = longestSuffix(word, ["iger", "igst", "ig"]);
  if (second === undefined || !inRegion(word, second, p1))
    return { word, changed: false };
  const prefix = word.slice(0, -second.length);
  if (prefix === "inn" || !isC(prefix)) return { word, changed: false };
  return { word: lengthenV(prefix), changed: true };
}

function validGeAt(word: string, index: number): boolean {
  const rest = word.slice(index + 2);
  if (rest.length < 3) return false;
  let cursor = 0;
  while (cursor < rest.length && vowelLengthAt(rest, cursor) === 0) cursor++;
  if (cursor >= rest.length) return false;
  while (cursor < rest.length) {
    const length = vowelLengthAt(rest, cursor);
    if (length === 0) break;
    cursor += length;
  }
  return cursor < rest.length;
}

function removeGe(word: string, index: number): string {
  let result = word.slice(0, index) + word.slice(index + 2);
  const char = result[index];
  if (char === "ë" || char === "ï")
    result = `${result.slice(0, index)}${char === "ë" ? "e" : "i"}${result.slice(index + 1)}`;
  return result;
}

function removePrefixGe(word: string): { word: string; changed: boolean } {
  if (!word.startsWith("ge") || !validGeAt(word, 0))
    return { word, changed: false };
  const rest = word.slice(2);
  if (
    rest.startsWith("eft") ||
    (rest.startsWith("val") && !rest.startsWith("vali")) ||
    rest.startsWith("vaa") ||
    rest.startsWith("vare")
  ) {
    return { word, changed: false };
  }
  return { word: removeGe(word, 0), changed: true };
}

function removeInfixGe(word: string): { word: string; changed: boolean } {
  const index = word.indexOf("ge", 1);
  return index >= 0 && validGeAt(word, index)
    ? { word: removeGe(word, index), changed: true }
    : { word, changed: false };
}

function step1c(word: string, p1: number): string {
  if (word.endsWith("d") && word.length - 1 >= p1 && isC(word.slice(0, -1))) {
    const prefix = word.slice(0, -1);
    if (prefix.endsWith("n") && prefix.length - 1 >= p1) return word;
    return prefix === "in" ? "inn" : prefix;
  }
  if (word.endsWith("t") && word.length - 1 >= p1 && isC(word.slice(0, -1))) {
    const prefix = word.slice(0, -1);
    if ((prefix.endsWith("h") && prefix.length - 1 >= p1) || prefix === "en")
      return word;
    return prefix;
  }
  return word;
}

function step7(word: string): { word: string; changed: boolean } {
  for (const suffix of ["kt", "ft", "pt"] as const) {
    if (word.endsWith(suffix))
      return { word: word.slice(0, -1), changed: true };
  }
  return { word, changed: false };
}

function step6(word: string): string {
  const pair = word.slice(-2);
  if (/^([bcdfghjklmnpqrstvwxyz])\1$/u.test(pair)) {
    if (pair === "nn" && word === "inn") return word;
    return word.slice(0, -1);
  }
  if (word.endsWith("v")) return `${word.slice(0, -1)}f`;
  if (word.endsWith("z")) return `${word.slice(0, -1)}s`;
  return word;
}

/** Snowball 3 Dutch stemmer, verified against snowball-data. */
export function stemDutch(word: string): string {
  if (word.length === 0) return word;
  let regions = measure(word);
  let result = word;
  let stemmed = false;
  for (const step of [
    () => step1(result, regions.p1),
    () => step2(result, regions.p1),
    () => step3(result, regions.p1, regions.p2),
    () => step4(result, regions.p1),
  ]) {
    const outcome = step();
    result = outcome.word;
    stemmed ||= outcome.changed;
  }
  const prefixGe = removePrefixGe(result);
  if (prefixGe.changed) {
    result = prefixGe.word;
    regions = measure(result);
    result = step1c(result, regions.p1);
    stemmed = true;
  }
  const infixGe = removeInfixGe(result);
  if (infixGe.changed) {
    result = infixGe.word;
    regions = measure(result);
    result = step1c(result, regions.p1);
    stemmed = true;
  }
  const seven = step7(result);
  result = seven.word;
  stemmed ||= seven.changed;
  return stemmed ? step6(result) : result;
}

File: python/searchable-analysis/src/searchable_analysis/stemmer_nl.py
import re

A = frozenset("aàáâä")
E = frozenset("eèéêë")
I = frozenset("iìíîï")
O = frozenset("oòóôö")
U = frozenset("uùúûü")
AEIOU = A | E | I | O | U
AIOU = A | I | O | U
VOWELS = AEIOU | {"y"}


def _vowel_length_at(word: str, index: int) -> int:
    if word[index:index + 2] == "ij":
        return 2
    return 1 if word[index:index + 1] in VOWELS else 0


def _measure(word: str) -> tuple[int, int]:
    regions: list[int] = []
    index = 0
    while len(regions) < 2 and index < len(word):
        while index < len(word) and _vowel_length_at(word, index) == 0:
            index += 1
        vowels = 0
        while index < len(word):
            length = _vowel_length_at(word, index)
            if length == 0:
                break
            vowels += 1
            index += length
        if vowels == 0 or index >= len(word):
            break
        index += 1
        regions.append(index)
    return (regions + [len(word), len(word)])[:2]


def _is_c(prefix: str) -> bool:
    return bool(prefix) and not prefix.endswith("ij") and prefix[-1] not in VOWELS


def _is_v(prefix: str) -> bool:
    return prefix.endswith("ij") or prefix[-1:] in VOWELS


def _in_region(word: str, suffix: str, region: int) -> bool:
    return len(word) - len(suffix) >= region


def _longest_suffix(word: str, suffixes: tuple[str, ...]) -> str | None:
    return max((suffix for suffix in suffixes if word.endswith(suffix)), key=len, default=None)


def _lengthen_v(word: str) -> str:
    if len(word) < 2 or word[-1] in VOWELS or word[-1] in "wx":
        return word
    last, prefix = word[-1], word[:-1]
    if prefix.endswith("eë"):
        return prefix[:-2] + "eëe" + last
    if prefix.endswith("ië"):
        return prefix[:-2] + "iee" + last
    vowel, before = prefix[-1:], prefix[:-1]
    if vowel in A | O | U:
        if not before or before[-1] not in AEIOU:
            return prefix + vowel + last
    elif vowel in E and vowel != "ë":
        if before and before[-1] in AEIOU:
            return word
        base = before if not before else before[:-1]
        if base[-1:] in AIOU or (len(base) == 1 and base in E):
            return word
        skipped = base[:-1]
        if len(skipped) >= 2 and skipped[-1] in AIOU and skipped[-2] not in AEIOU:
            return word
        return prefix + vowel + last
    return word


def _step1(word: str, p1: int) -> tuple[str, bool]:
    suffix = _longest_suffix(word, ("'s", "ies", "és", "aus", "nde", "en", "es", "s"))
    if suffix is None:
        return word, False
    prefix = word[:-len(suffix)]
    if suffix == "'s":
        return prefix, True
    if suffix == "s":
        if not _in_region(word, suffix, p1) or not _is_c(prefix) or (prefix.endswith("t") and len(prefix) - 1 >= p1):
            return word, False
        return prefix, True
    if suffix == "ies":
        return (prefix + "ie", True) if _in_region(word, suffix, p1) else (word, False)
    if suffix == "es":
        if prefix.endswith("ar") and len(prefix) - 2 >= p1 and _is_c(prefix[:-2]):
            return _lengthen_v(prefix), True
        if prefix.endswith("er") and len(prefix) - 2 >= p1 and _is_c(prefix[:-2]):
            return prefix, True
        return (prefix + "e", True) if _in_region(word, suffix, p1) and _is_c(prefix) else (word, False)
    if suffix == "és":
        return (prefix + "é", True) if _in_region(word, suffix, p1) else (word, False)
    if suffix == "aus":
        return (prefix + "au", True) if _in_region(word, suffix, p1) and _is_v(prefix) else (word, False)
    if suffix == "en":
        if prefix.endswith("hed") and len(prefix) - 3 >= p1:
            return prefix[:-3] + "heid", True
        if prefix.endswith("nd"):
            return prefix, True
        if prefix.endswith("d") and len(prefix) - 1 >= p1 and _is_c(prefix[:-1]):
            return prefix[:-1], True
        if prefix.endswith(("i", "j")) and _is_v(prefix[:-1]):
            return prefix, True
        return (_lengthen_v(prefix), True) if _in_region(word, suffix, p1) and _is_c(prefix) else (word, False)
    if suffix == "nde":
        return prefix + "nd", True
    return word, False


def _step2(word: str, p1: int) -> tuple[str, bool]:
    suffix = _longest_suffix(word, ("lijke", "ische", "ieve", "ene", "je", "ge", "de", "te", "se", "re", "le"))
    if suffix is None:
        return word, False
    prefix = word[:-len(suffix)]
    if suffix == "je":
        if prefix.endswith("'t"):
            return prefix[:-2], True
        if prefix.endswith("et") and len(prefix) - 2 >= p1 and _is_c(prefix[:-2]):
            return prefix[:-2], True
        if prefix.endswith("rnt"):
            return prefix[:-3] + "rn", True
        if prefix.endswith("t") and len(prefix) - 1 >= p1 and _is_v(prefix[:-2]):
            return prefix[:-1], True
        if prefix.endswith("ink"):
            return prefix[:-3] + "ing", True
        if prefix.endswith("mp"):
            return prefix[:-2] + "m", True
        if prefix.endswith("'") and len(prefix) - 1 >= p1:
            return prefix[:-1], True
        return (prefix, True) if _in_region(word, suffix, p1) and _is_c(prefix) else (word, False)
    if not _in_region(word, suffix, p1):
        return word, False
    replacements = {"ge": "g", "lijke": "lijk", "ische": "isch", "te": "t", "se": "s", "re": "r"}
    if suffix in replacements:
        return prefix + replacements[suffix], True
    if suffix == "de":
        return (prefix, True) if _is_c(prefix) else (word, False)
    if suffix == "le":
        return _lengthen_v(prefix + "l"), True
    if suffix == "ene":
        return (_lengthen_v(prefix + "en"), True) if _is_c(prefix) else (word, False)
    if suffix == "ieve":
        return (prefix + "ief", True) if _is_c(prefix) else (word, False)
    return word, False


def _step3(word: str, p1: int, p2: int) -> tuple[str, bool]:
    suffix = _longest_suffix(word, ("iteit", "isme", "erij", "arij", "heid", "ster", "rder", "atie", "ing", "sel", "fie", "gie", "tst", "dst"))
    if suffix is None:
        return word, False
    prefix = word[:-len(suffix)]
    if suffix == "atie":
        return (prefix + "eer", True) if _in_region(word, suffix, p1) else (word, False)
    if suffix == "iteit":
        return (_lengthen_v(prefix), True) if _in_region(word, suffix, p1) else (word, False)
    if suffix in ("heid", "sel", "ster"):
        return (prefix, True) if _in_region(word, suffix, p1) else (word, False)
    if suffix == "rder":
        return prefix + "r", True
    if suffix in ("ing", "isme", "erij"):
        if prefix.endswith("ild"):
            return prefix + "er", True
        return (_lengthen_v(prefix), True) if _in_region(word, suffix, p1) else (word, False)
    if suffix == "arij":
        return (prefix + "aar", True) if _in_region(word, suffix, p1) and _is_c(prefix) else (word, False)
    if suffix in ("fie", "gie"):
        return (_lengthen_v(prefix + suffix[0]), True) if _in_region(word, suffix, p2) else (word, False)
    if suffix in ("tst", "dst"):
        return (prefix + suffix[0], True) if _in_region(word, suffix, p1) and _is_c(prefix) else (word, False)
    return word, False


def _step4(word: str, p1: int) -> tuple[str, bool]:
    first = _longest_suffix(word, ("achtiger", "achtigst", "eriger", "erigst", "ioneel", "lijker", "lijkst", "achtig", "atief", "baar", "naar", "laar", "raar", "tant", "erig", "end"))
    if first is not None and _in_region(word, first, p1):
        prefix = word[:-len(first)]
        replacements = {"ioneel": "ie", "atief": "eer", "tant": "teer", "lijker": "lijk", "lijkst": "lijk"}
        if first in replacements:
            return prefix + replacements[first], True
        if first == "baar" or first.startswith("achtig"):
            return prefix, True
        if first in ("naar", "laar", "raar") and _is_v(prefix):
            return prefix + first[0], True
        if first in ("erig", "eriger", "erigst", "end") and _is_c(prefix):
            return _lengthen_v(prefix), True
    second = _longest_suffix(word, ("iger", "igst", "ig"))
    if second is None or not _in_region(word, second, p1):
        return word, False
    prefix = word[:-len(second)]
    if prefix == "inn" or not _is_c(prefix):
        return word, False
    return _lengthen_v(prefix), True


def _valid_ge_at(word: str, index: int) -> bool:
    rest = word[index + 2:]
    if len(rest) < 3:
        return False
    cursor = 0
    while cursor < len(rest) and _vowel_length_at(rest, cursor) == 0:
        cursor += 1
    if cursor >= len(rest):
        return False
    while cursor < len(rest):
        length = _vowel_length_at(rest, cursor)
        if length == 0:
            break
        cursor += length
    return cursor < len(rest)


def _remove_ge(word: str, index: int) -> str:
    result = word[:index] + word[index + 2:]
    if result[index:index + 1] in ("ë", "ï"):
        result = result[:index] + ("e" if result[index] == "ë" else "i") + result[index + 1:]
    return result


def _remove_prefix_ge(word: str) -> tuple[str, bool]:
    if not word.startswith("ge") or not _valid_ge_at(word, 0):
        return word, False
    rest = word[2:]
    if rest.startswith("eft") or (rest.startswith("val") and not rest.startswith("vali")) or rest.startswith(("vaa", "vare")):
        return word, False
    return _remove_ge(word, 0), True


def _remove_infix_ge(word: str) -> tuple[str, bool]:
    index = word.find("ge", 1)
    return (_remove_ge(word, index), True) if index >= 0 and _valid_ge_at(word, index) else (word, False)


def _step1c(word: str, p1: int) -> str:
    if word.endswith("d") and len(word) - 1 >= p1 and _is_c(word[:-1]):
        prefix = word[:-1]
        if prefix.endswith("n") and len(prefix) - 1 >= p1:
            return word
        return "inn" if prefix == "in" else prefix
    if word.endswith("t") and len(word) - 1 >= p1 and _is_c(word[:-1]):
        prefix = word[:-1]
        if (prefix.endswith("h") and len(prefix) - 1 >= p1) or prefix == "en":
            return word
        return prefix
    return word


def _step7(word: str) -> tuple[str, bool]:
    return (word[:-1], True) if word.endswith(("kt", "ft", "pt")) else (word, False)


def _step6(word: str) -> str:
    pair = word[-2:]
    if re.fullmatch(r"([bcdfghjklmnpqrstvwxyz])\1", pair):
        return word if pair == "nn" and word == "inn" else word[:-1]
    if word.endswith("v"):
        return word[:-1] + "f"
    if word.endswith("z"):
        return word[:-1] + "s"
    return word


def stem_dutch(word: str) -> str:
    if not word:
        return word
    p1, p2 = _measure(word)
    result, stemmed = word, False
    for step in (_step1, _step2):
        result, changed = step(result, p1)
        stemmed |= changed
    result, changed = _step3(result, p1, p2)
    stemmed |= changed
    result, changed = _step4(result, p1)
    stemmed |= changed
    result, changed = _remove_prefix_ge(result)
    if changed:
        p1, _ = _measure(result)
        result = _step1c(result, p1)
        stemmed = True
    result, changed = _remove_infix_ge(result)
    if changed:
        p1, _ = _measure(result)
        result = _step1c(result, p1)
        stemmed = True
    result, changed = _step7(result)
    stemmed |= changed
    return _step6(result) if stemmed else result
